src/odm_sharing/private/queries.py

```diff
@@ -1,14 +1,326 @@
-from dataclasses import dataclass
-from typing import Dict
+import re
+from collections import defaultdict
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Dict, List, Tuple
+# from pprint import pprint
 
+from functional import seq
+
+from odm_sharing.private.common import OrgName, TableName
 from odm_sharing.private.rules import RuleId
+from odm_sharing.private.stdext import StrEnum, sorted_dict
+from odm_sharing.private.utils import dqt
+
+from odm_sharing.private.trees import (
+    ALL_LIT,
+    Node,
+    NodeKind,
+    Op,
+    RangeKind,
+    RuleTree,
+    parse_op,
+)
+
+
+Args = List[str]
+Sql = str
 
 
-SqlQuery = str
+class SqlDialect(StrEnum):
+    OTHER = ''
+    MSSQL = 'mssql'
+    SYBASE = 'sybase'
 
 
 @dataclass(frozen=True)
 class Query:
-    data_sql: SqlQuery
-    rule_count_sqls: Dict[RuleId, SqlQuery]
-    column_sql: SqlQuery
+    sql: Sql
+    args: Args = field(default_factory=list)
+
+
+PartialQuery = Query  # incomplete query
+
+
+@dataclass(frozen=True)
+class TableQuery:
+    '''collection of queries for a single table'''
+    data_query: Query
+    rule_count_queries: Dict[RuleId, Query]
+    columns: List[str]  # pre-specified columns
+    select_rule_id: RuleId
+    _select_query: Query
+
+
+OrgTableQueries = Dict[OrgName, Dict[TableName, TableQuery]]
+
+
+INVALID_IDENT_PATTERN = re.compile(r'\W+', re.ASCII)
+
+
+def ident(x: str) -> str:
```

Comment:
identifier, as I wrote in the docstring, meaning a name that identifies something (like a column)

Reply:
Can we rename to `sanitize_ident`?

Reply:
that's a very long name, and it implies that it only does sanitation but it's actually quoting as well. I think `ident` was a good name, but I can do `sql_ident` if that helps? It both formats and makes sure to avoid injections.

Reply:
actually, even `sql_ident` is too long for my taste. I was initially using `dqt` (double quote) to quote the identifiers, but then I needed sanitation as well. The need for keeping it so short comes from the way it's used, in formatted strings, like `aasd as das {ident(d)} asfsd {ident(asd)}`, which quickly become too long.

Reply:
resolved in chat